crates/cxx-qt-lib-extras/src/gui/qapplication.rs

```diff
@@ -93,6 +110,24 @@ mod ffi {
 
 pub use ffi::QApplication;
 
+impl Upcast<QGuiApplication> for QApplication {
+    unsafe fn upcast_ptr(this: *const Self) -> *const QGuiApplication {
+        ffi::upcast_qapplication(this)
+    }
+
+    unsafe fn from_base_ptr(base: *const QGuiApplication) -> *const Self {
+        ffi::downcast_qapplication(base)
+    }
+}
+
+impl Deref for QApplication {
```

Comment:
I think it was discussed to do Deref automatically for Upcast before, but it was decided against, @LeonMatthesKDAB @BenFordTytherington can you remember the details?

Reply:
The reason we don't automatically support this is because of two reasons:

1. We already implement Deref for generated QObjects. They Deref to the inner Rust type.
2. There could be multiple-inheritance on the C++ side, in which case it would be unclear which of the base classes to deref to (though this is currently academic, we don't support it anyway).

For types like QApplication, none of these apply, so we can absolutely `impl Deref` via Upcasting.